[documentation]: add wasm README (#1919)
hyperledger/iroha: wasm/README.md

# How to optimize your binary size?

> Smartcontracts should be optimized for because they are stored directly on the blockchain. By

Reviewer: ```suggestion
Smartcontracts should be optimized, because they are stored directly on the blockchain. By
```

following this list of optimization steps your binary's size can be reduced by an order of magnitude
(e.g. from 1.1MB to 100KB):

> 1. Create a `Cargo.toml` akin to the following template:

Reviewer: ```suggestion
1. Create a `Cargo.toml` following this template: 
```

```toml
  [package]
  name = "smartcontract"
  version = "0.1.0"
  edition = "2021"

  [lib]

>   crate-type = ['cdylib']    # Crate has to be linked dynamically

Reviewer: Add explanation. 

A user knows that Rust is statically linked by default because of unstable ABI, meaning that binaries compiled by different compilers might not be compatible. 

The reason why we want dynamic linking, is because we will store wasm VMs compatible with all binary versions anyway, so the dynamic libraries will always be available. This is a contract that we should make explicitly with the developers of smartcontracts.

  [profile.release]
  strip = "debug"     # Remove debugging info from the binary
  panic = "abort"     # Panics are transcribed to Traps when compiling for wasm anyways
  lto = true          # Link-time-optimization produces notable decrease in binary size
  opt-level = "z"     # Optimize for size vs speed with "s"/"z"(removes vectorization)
  codegen-units = 1   # Further reduces binary size but increases compilation time

  [dependencies]

>   iroha_data_model = { git = "https://github.com/hyperledger/iroha/", branch = "iroha2", default-features = false }

Reviewer: Do we still need `no_std` if we do both LTO and dynamically link the smartcontracts?

Author: yes, it seems so. I think that is because LTO happens after successful compilation of modules and it's not possible to compile `std` for `wasm32` target

  iroha_wasm = { git = "https://github.com/hyperledger/iroha/", branch = "iroha2" }
```

2. Re-build `libcore` and `alloc` with excluded panicking infrastructure
```
  cargo +nightly build -Z build-std -Z build-std-features=panic_immediate_abort --target wasm32-unknown-unknown
```
**NOTE**: This cargo feature is unstable and may not be suitable for production

3. Use [wasm-opt](https://github.com/WebAssembly/binaryen) to further optimize the built binary:
```sh
$ wasm-opt -Os -o output.wasm input.wasm
```

> While these settings will certainly result in a whooping size reduction of a binary,

Reviewer: ```suggestion
Following these steps is the bare minimum that can be done to all WASM smartcontracts. Users are encouraged to further reduce the sizes of their binaries by profiling using [twiggy](https://rustwasm.github.io/twiggy/) and avoid usage of large libraries. 
```
